test/indexes.test-d.ts

```diff
@@ -259,3 +259,271 @@ async function main() {
+      include2?: boolean | undefined;
+      include3?: number | undefined;
+    }[]
+  >(resp.data);
+});
+
+// scanning index with projected attributes should only allow filtering on projected attributes
+projectionEntity.scan.includeIndex
+  .where((attrs, ops) => {
+    expectError(() => ops.eq(attrs.id, "1"));
+    return ops.eq(attrs.include1, "abc");
+  })
+  .go()
+  .then((resp) => {
+    expectType<
+      {
+        include1: string;
+        include2?: boolean | undefined;
+        include3?: number | undefined;
+      }[]
+    >(resp.data);
+  });
+
+// scanning index with projected attributes with hydrate should only allow filtering on projected attributes
+projectionEntity.scan.includeIndex
+  .where((attrs, ops) => {
+    expectError(() => ops.eq(attrs.id, "1"));
+    return ops.eq(attrs.include1, "abc");
+  })
+  .where((attrs, ops) => ops.eq(attrs.include2, true))
+  .go({ hydrate: true })
+  .then((resp) => {
+    expectType<
+      {
+        id: string;
+        include1: string;
+        include2?: boolean | undefined;
+        include3?: number | undefined;
+        exclude1?: string | undefined;
+        exclude2?: number | undefined;
+        exclude3?: boolean | undefined;
+      }[]
+    >(resp.data);
+  });
+
+// scanning index with projected attributes with hydrate should return user-provided attributes but not allow filtering on non-projected attributes
+projectionEntity.scan.includeIndex
+  .where((attrs, ops) => {
+    expectError(() => ops.eq(attrs.id, "1"));
+    return ops.eq(attrs.include1, "abc");
+  })
+  .where((attrs, ops) => ops.eq(attrs.include2, true))
+  .go({ hydrate: true, attributes: ["id"] })
+  .then((resp) => {
+    expectType<
+      {
+        id: string;
+      }[]
+    >(resp.data);
+  });
+
+// scanning index with projected items should not allow to specify non-projected attributes
+expectError(() =>
+  projectionEntity.scan.includeIndex.go({ attributes: ["id"] }),
+);
+
+// scanning index with projected items and hydrate should allow to specify non-projected attributes
+projectionEntity.scan.includeIndex
+  .where((attrs, ops) => ops.eq(attrs.include2, true))
+  .go({ hydrate: true, attributes: ["id", "include1"] })
+  .then((resp) => {
+    expectType<
+      {
+        id: string;
+        include1: string;
+      }[]
+    >(resp.data);
+  });
+
+// ==== QUERYING ====
+
+// querying index with projected attributes should only return the projected attributes
+projectionEntity.query
+  .includeIndex({ id: "1" })
+  .go()
+  .then((resp) => {
+    expectType<
+      {
+        include1: string;
+        include2?: boolean | undefined;
+        include3?: number | undefined;
+      }[]
+    >(resp.data);
+  });
+
+// querying index with projected attributes should only allow filtering on projected attributes
+projectionEntity.query
+  .includeIndex({ id: "1" })
+  .where((attrs, ops) => {
+    expectError(() => ops.eq(attrs.id, "1"));
+    return ops.eq(attrs.include1, "abc");
```

Comment:
(This applies the `where` callbacks below as well) It looks like the attribute value typing on `attr` is dropped for `query` and `scan`. For example, if you add `expectError(() => ops.eq(attrs.include1, 123));` it doesn't find an error.

Reply:
@tywalch amazing catch! Fixed in my latest commit and added tests for this as well.